### PR TITLE
Fix Context Parallel NaN Loss

### LINKED ISSUE #1764: [BUG] Context parallel `nan` loss
**Describe the bug**

When enabling context parallel and when all the tokens on a single GPU is masked out, the loss and grad becomes NaN.

see also: https://github.com/NVIDIA/NeMo/issues/14488

**To Reproduce**
Use CP > 2 and pad a short training data 2x it's original length 

**Expected behavior**
model trains.

**Environment (please complete the following information):**
 - Megatron-LM commit ID version 0.31.1
 - PyTorch version
 - CUDA version
 - NCCL version

**Proposed fix**
I've identified the issue. submitting a PR soon

### PR DESCRIPTION
**Pull request by leoleoasd**
fixes #1764 and https://github.com/NVIDIA/NeMo/issues/14488

### CODE REVIEW

**Comment by leoleoasd** [comment]
@sbhavani Is there any update on this?

**Comment by sbhavani** [comment]
@leoleoasd thanks for reporting the issue and providing the fix!

I see that we recently added commit [1aef9f8ef](https://github.com/NVIDIA/Megatron-LM/commit/1aef9f8ef) which provides protection against division by zero. I'll review internally and check which approach is appropriate for the loss calculation.

**Comment by sbhavani** [comment]
@leoleoasd is your issue resolved with the recent commit?

**Comment by leoleoasd** [comment]
I didn't get a chance to try it, will find sometime next week
